File: c++/source/model/lorenz63.cpp

/*__________________________________________________
 * model/
 * lorenz63.cpp
 *__________________________________________________
 * author        : colonel
 * last modified : 2016/9/16
 *__________________________________________________
 *
 * implementation of class Lorenz63Model
 *
 */

#ifndef MODEL_LORENZ63_CPP
#define MODEL_LORENZ63_CPP

#include "lorenz63.h"

namespace model
{

    // constructor
    template <typename Real, template <std::size_t, typename> class Integrator, template <std::size_t, typename> class ModelErrorGenerator>
        Lorenz63Model <Real, Integrator, ModelErrorGenerator> ::
        Lorenz63Model() :
            m_sigma(10.0),
            m_rho(28.0),
            m_beta(2.667),
            m_meg(),
            m_integrator()
        {
        }

    // set model parameters
    template <typename Real, template <std::size_t, typename> class Integrator, template <std::size_t, typename> class ModelErrorGenerator>
        void Lorenz63Model <Real, Integrator, ModelErrorGenerator> ::
        setModelParameters(Real t_sigma, Real t_rho, Real t_beta)
        {
            m_sigma = t_sigma ;
            m_rho   = t_rho ;
            m_beta  = t_beta ;
        }

    // access function for the model error generator
    // used to set the rng and the model error parameters
    template <typename Real, template <std::size_t, typename> class Integrator, template <std::size_t, typename> class ModelErrorGenerator>
        ModelErrorGenerator <3, Real> & Lorenz63Model <Real, Integrator, ModelErrorGenerator> ::
        modelErrorGenerator()
        {
            return m_meg ;
        }

    // step forward
    // t_stateNPP = Model(t_stateN) + error
    // computed using the integration method
    template <typename Real, template <std::size_t, typename> class Integrator, template <std::size_t, typename> class ModelErrorGenerator>
        void Lorenz63Model <Real, Integrator, ModelErrorGenerator> ::
        forward(Real t_dt, const Real * t_stateN, Real * t_stateNPP, bool addModelError)
        {
            m_integrator.forward(t_dt, t_stateN, t_stateNPP, *this) ;

            if(addModelError)
            {
                m_meg.addError(t_stateNPP) ;
            }
        }

    // compute dx according to the model
    template <typename Real, template <std::size_t, typename> class Integrator, template <std::size_t, typename> class ModelErrorGenerator>
        void Lorenz63Model <Real, Integrator, ModelErrorGenerator> ::
        computeDerivative(const Real * t_state, Real * t_dstate)
        {
            t_dstate[0] = m_sigma * ( t_state[1] - t_state[0] ) ;
            t_dstate[1] = ( m_rho - t_state[2] ) * t_state[0] - t_state[1] ;
            t_dstate[2] = t_state[0] * t_state[1] - m_beta * t_state[2] ;
        }

}

#endif // MODEL_LORENZ63_CPP

File: python/pyLorenz/model/lorenz63.py
#! /usr/bin/env python

#__________________________________________________
# pyLorenz/model/
# lorenz63.py
#__________________________________________________
# author        : colonel
# last modified : 2016/10/13
#__________________________________________________
#
# classes to handle the a Lorenz 1963 model
# i.e. process compute the dervative at a given point according to a Lorenz 1963 model
#

#__________________________________________________

class Lorenz63Model(object):

    #_________________________

    def __init__(self, t_sigma = 10.0, t_beta = 8.0 / 3.0, t_rho = 28.0):
        self.setLorenz63ModelParameters(t_sigma, t_beta, t_rho)

    #_________________________

    def setLorenz63ModelParameters(self, t_sigma, t_beta, t_rho):
        # state space dimension
        self.m_spaceDimension = 3
        # model parameters
        self.m_sigma          = t_sigma
        self.m_beta           = t_beta
        self.m_rho            = t_rho

    #_________________________

    def __call__(self, t_x, t_t, t_dx):
        # compute dx at point x and time t according to Lorenz 1963 model
        t_dx[..., 0] = self.m_sigma * ( t_x[..., 1] - t_x[..., 0] )
        t_dx[..., 1] = ( self.m_rho - t_x[..., 2] ) * t_x[..., 0] - t_x[..., 1]
        t_dx[..., 2] = t_x[..., 0] * t_x[..., 1] - self.m_beta * t_x[..., 2]

#__________________________________________________

